File: gcs/macos/Sources/GCSApp/RosbridgeClient.swift
import Foundation

/// Minimal rosbridge v2 client over a WebSocket (ws://host:9090).
///
/// Subscribes to the six telemetry topics the dashboard needs and pushes parsed
/// values into a TelemetryStore on the main thread. Auto-reconnects on drop.
/// The GCS is monitor-only, so this client never publishes or calls services.
final class RosbridgeClient: NSObject {
    private let store: TelemetryStore
    private var task: URLSessionWebSocketTask?
    private var session: URLSession!
    private var host = ""
    private var active = false
    private var generation = 0

    private struct Sub { let topic: String; let type: String }
    private let subs: [Sub] = [
        .init(topic: "/mavros/state",                    type: "mavros_msgs/msg/State"),
        .init(topic: "/mavros/statustext/recv",          type: "mavros_msgs/msg/StatusText"),
        .init(topic: "/mavros/battery",                   type: "sensor_msgs/msg/BatteryState"),
        .init(topic: "/mavros/mavros/pose",               type: "geometry_msgs/msg/PoseStamped"),
        .init(topic: "/mavros/mavros/velocity_local",     type: "geometry_msgs/msg/TwistStamped"),
        .init(topic: "/mavros/mavros/data",               type: "sensor_msgs/msg/Imu"),
    ]

    init(store: TelemetryStore) {
        self.store = store
        super.init()
        session = URLSession(configuration: .ephemeral)
    }

    func connect(host: String) {
        disconnect()
        self.host = host
        active = true
        generation += 1
        openSocket(generation)
    }

    func disconnect() {
        active = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        DispatchQueue.main.async { self.store.resetForReconnect() }
    }

    private func openSocket(_ gen: Int) {
        guard active, gen == generation, let url = URL(string: "ws://\(host):9090") else { return }
        let t = session.webSocketTask(with: url)
        task = t
        t.resume()
        // rosbridge has no explicit "ready" handshake; subscribe immediately.
        for s in subs { send(subscribe: s) }
        DispatchQueue.main.async {
            self.store.rosConnected = true
            self.store.log("rosbridge connected (\(self.host))")
        }
        receive(gen)
    }

    private func send(subscribe s: Sub) {
        let msg: [String: Any] = ["op": "subscribe", "topic": s.topic, "type": s.type]
        guard let data = try? JSONSerialization.data(withJSONObject: msg),
              let str = String(data: data, encoding: .utf8) else { return }
        task?.send(.string(str)) { _ in }
    }

    private func receive(_ gen: Int) {
        task?.receive { [weak self] result in
            guard let self = self, gen == self.generation else { return }
            switch result {
            case .failure:
                self.handleDrop(gen)
            case .success(let message):
                switch message {
                case .string(let s): self.handle(text: s)
                case .data(let d): self.handle(text: String(decoding: d, as: UTF8.self))
                @unknown default: break
                }
                self.receive(gen)
            }
        }
    }

    private func handleDrop(_ gen: Int) {
        guard active, gen == generation else { return }
        DispatchQueue.main.async {
            self.store.rosConnected = false
            self.store.log("rosbridge lost — reconnecting…")
        }
        // Reconnect after 3 s (matches web behaviour).
        DispatchQueue.global().asyncAfter(deadline: .now() + 3) { [weak self] in
            guard let self = self, self.active, gen == self.generation else { return }
            self.openSocket(gen)
        }
    }

    // MARK: - Message parsing

    private func handle(text: String) {
        guard let data = text.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              (obj["op"] as? String) == "publish",
              let topic = obj["topic"] as? String,
              let msg = obj["msg"] as? [String: Any] else { return }

        DispatchQueue.main.async { self.apply(topic: topic, msg: msg) }
    }

    private func apply(topic: String, msg: [String: Any]) {
        switch topic {
        case "/mavros/state":
            if let mode = msg["mode"] as? String, mode != store.mode {
                store.mode = mode
                store.log("mode → \(mode)")
            }
            if let armed = msg["armed"] as? Bool { store.armed = armed }
            store.touched()

        case "/mavros/statustext/recv":
            let sev = (msg["severity"] as? Int) ?? 99
            if sev <= 4, let txt = msg["text"] as? String { store.log("[FCU] \(txt)") }

        case "/mavros/battery":
            let v = (msg["voltage"] as? Double) ?? Double.nan
            let pRaw = (msg["percentage"] as? Double) ?? Double.nan
            if v.isFinite && v > 0 && v < 60 { store.batteryVoltage = v } else { store.batteryVoltage = nil }
            var pct: Double? = nil
            if pRaw.isFinite && pRaw >= 0 && pRaw <= 1 {
                pct = pRaw * 100
            } else if let v = store.batteryVoltage {
                pct = max(0, min(100, (v / 6 - 3.3) / (4.2 - 3.3) * 100)) // 6S estimate
            }
            store.batteryPct = pct
            if let pct = pct {
                store.pushChart(\.chartBat, pct)
                if pct < 15, let v = store.batteryVoltage {
                    store.log(String(format: "WARNING LOW BATTERY: %.2f V", v))
                }
            }
            store.touched()

        case "/mavros/mavros/pose":
            guard let pose = msg["pose"] as? [String: Any],
                  let p = pose["position"] as? [String: Any],
                  let o = pose["orientation"] as? [String: Any] else { return }
            store.posX = (p["x"] as? Double) ?? 0
            store.posY = (p["y"] as? Double) ?? 0
            store.posZ = (p["z"] as? Double) ?? 0
            let e = quatToEuler((o["x"] as? Double) ?? 0, (o["y"] as? Double) ?? 0,
                                (o["z"] as? Double) ?? 0, (o["w"] as? Double) ?? 1)
            store.roll = e.roll; store.pitch = e.pitch; store.yaw = e.yaw
            store.pushTrail(store.posX, store.posY)
            store.pushChart(\.chartAlt, store.posZ)
            store.touched()

        case "/mavros/mavros/velocity_local":
            guard let twist = msg["twist"] as? [String: Any],
                  let l = twist["linear"] as? [String: Any] else { return }
            store.vx = (l["x"] as? Double) ?? 0
            store.vy = (l["y"] as? Double) ?? 0
            store.vz = (l["z"] as? Double) ?? 0
            store.pushChart(\.chartSpd, store.speedH)
            store.touched()

        case "/mavros/mavros/data":
            guard let av = msg["angular_velocity"] as? [String: Any],
                  let la = msg["linear_acceleration"] as? [String: Any] else { return }
            let gx = (av["x"] as? Double) ?? 0, gy = (av["y"] as? Double) ?? 0, gz = (av["z"] as? Double) ?? 0
            let ax = (la["x"] as? Double) ?? 0, ay = (la["y"] as? Double) ?? 0, az = (la["z"] as? Double) ?? 0
            store.gyroMag = (gx * gx + gy * gy + gz * gz).squareRoot() * 180 / Double.pi
            store.accelMag = (ax * ax + ay * ay + az * az).squareRoot()

        default: break
        }
    }
}

File: gcs/windows/src/BumblebeeGcs/Services/RosbridgeClient.cs
// Copyright 2026 FutureLab
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using BumblebeeGcs.Models;
using Microsoft.UI.Dispatching;

namespace BumblebeeGcs.Services;

/// <summary>
/// Минимальный rosbridge v2 клиент по WebSocket (ws://host:9090).
/// Подписывается на шесть топиков телеметрии и толкает распарсенные значения
/// в TelemetryStore на UI-потоке. Авто-реконнект при обрыве через 3 с.
/// GCS работает в режиме монитора — клиент никогда не публикует и не зовёт сервисы.
/// </summary>
public sealed class RosbridgeClient
{
    private readonly TelemetryStore _store;
    private readonly DispatcherQueue _dq;
    private ClientWebSocket? _ws;
    private string _host = "";
    private bool _active;
    private int _generation;

    private readonly record struct Sub(string Topic, string Type);
    private static readonly Sub[] Subs =
    {
        new("/mavros/state",                "mavros_msgs/msg/State"),
        new("/mavros/statustext/recv",      "mavros_msgs/msg/StatusText"),
        new("/mavros/battery",              "sensor_msgs/msg/BatteryState"),
        new("/mavros/mavros/pose",          "geometry_msgs/msg/PoseStamped"),
        new("/mavros/mavros/velocity_local","geometry_msgs/msg/TwistStamped"),
        new("/mavros/mavros/data",          "sensor_msgs/msg/Imu"),
    };

    public RosbridgeClient(TelemetryStore store, DispatcherQueue dq)
    {
        _store = store;
        _dq = dq;
    }

    public void Connect(string host)
    {
        Disconnect();
        _host = host;
        _active = true;
        var gen = ++_generation;
        _ = OpenSocketAsync(gen);
    }

    public void Disconnect()
    {
        _active = false;
        _generation++;
        try { _ws?.Abort(); } catch { }
        _ws = null;
        _dq.TryEnqueue(_store.ResetForReconnect);
    }

    private async Task OpenSocketAsync(int gen)
    {
        if (!_active || gen != _generation) return;
        var ws = new ClientWebSocket();
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
            await ws.ConnectAsync(new Uri($"ws://{_host}:9090"), cts.Token);
        }
        catch
        {
            ws.Dispose();
            HandleDrop(gen);
            return;
        }
        if (!_active || gen != _generation) { ws.Abort(); ws.Dispose(); return; }
        _ws = ws;

        foreach (var s in Subs)
        {
            var msg = JsonSerializer.Serialize(new { op = "subscribe", topic = s.Topic, type = s.Type });
            try
            {
                await ws.SendAsync(Encoding.UTF8.GetBytes(msg), WebSocketMessageType.Text, true, CancellationToken.None);
            }
            catch { HandleDrop(gen); return; }
        }

        _dq.TryEnqueue(() =>
        {
            _store.SetRosConnected(true);
            _store.Log($"rosbridge connected ({_host})");
            _store.NotifyUpdated();
        });

        await ReceiveLoopAsync(ws, gen);
    }

    private async Task ReceiveLoopAsync(ClientWebSocket ws, int gen)
    {
        var chunk = new byte[64 * 1024];
        var message = new MemoryStream();
        while (_active && gen == _generation)
        {
            WebSocketReceiveResult result;
            try { result = await ws.ReceiveAsync(chunk, CancellationToken.None); }
            catch { HandleDrop(gen); return; }
            if (result.MessageType == WebSocketMessageType.Close) { HandleDrop(gen); return; }
            message.Write(chunk, 0, result.Count);
            if (!result.EndOfMessage) continue;
            HandleText(Encoding.UTF8.GetString(message.GetBuffer(), 0, (int)message.Length));
            message.SetLength(0);
        }
    }

    private void HandleDrop(int gen)
    {
        if (!_active || gen != _generation) return;
        _dq.TryEnqueue(() =>
        {
            _store.SetRosConnected(false);
            _store.Log("rosbridge lost — reconnecting…");
            _store.NotifyUpdated();
        });
        // Реконнект через 3 с (совпадает с поведением веба/мака).
        _ = Task.Delay(3000).ContinueWith(_ =>
        {
            if (_active && gen == _generation) _ = OpenSocketAsync(gen);
        });
    }

    // MARK: - Message parsing

    private void HandleText(string text)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(text); }
        catch { return; }
        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return;
            if (!root.TryGetProperty("op", out var op) || op.GetString() != "publish") return;
            if (!root.TryGetProperty("topic", out var topicEl) || topicEl.GetString() is not string topic) return;
            if (!root.TryGetProperty("msg", out var msg) || msg.ValueKind != JsonValueKind.Object) return;
            // JsonElement живёт внутри документа — клонируем для перехода на UI-поток.
            var msgClone = msg.Clone();
            _dq.TryEnqueue(() => Apply(topic, msgClone));
        }
    }

    static double? Num(JsonElement e, string prop) =>
        e.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.Number ? v.GetDouble() : null;

    static JsonElement? Obj(JsonElement e, string prop) =>
        e.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.Object ? v : null;

    private void Apply(string topic, JsonElement msg)
    {
        switch (topic)
        {
            case "/mavros/state":
            {
                if (msg.TryGetProperty("mode", out var modeEl) && modeEl.GetString() is string mode && mode != _store.Mode)
                {
                    _store.Mode = mode;
                    _store.Log($"mode → {mode}");
                }
                if (msg.TryGetProperty("armed", out var armedEl) &&
                    armedEl.ValueKind is JsonValueKind.True or JsonValueKind.False)
                    _store.Armed = armedEl.GetBoolean();
                _store.Touch();
                break;
            }
            case "/mavros/statustext/recv":
            {
                var sev = (int)(Num(msg, "severity") ?? 99);
                if (sev <= 4 && msg.TryGetProperty("text", out var txtEl) && txtEl.GetString() is string txt)
                    _store.Log($"[FCU] {txt}");
                break;
            }
            case "/mavros/battery":
            {
                var v = Num(msg, "voltage") ?? double.NaN;
                var pRaw = Num(msg, "percentage") ?? double.NaN;
                _store.BatteryVoltage = double.IsFinite(v) && v > 0 && v < 60 ? v : null;
                double? pct = null;
                if (double.IsFinite(pRaw) && pRaw >= 0 && pRaw <= 1)
                    pct = pRaw * 100;
                else if (_store.BatteryVoltage is double bv)
                    pct = Math.Max(0, Math.Min(100, (bv / 6 - 3.3) / (4.2 - 3.3) * 100)); // 6S estimate
                _store.BatteryPct = pct;
                if (pct is double p)
                {
                    TelemetryStore.PushChart(_store.ChartBat, p);
                    if (p < 15 && _store.BatteryVoltage is double bv2)
                        _store.Log($"WARNING LOW BATTERY: {bv2:F2} V");
                }
                _store.Touch();
                break;
            }
            case "/mavros/mavros/pose":
            {
                if (Obj(msg, "pose") is not JsonElement pose ||
                    Obj(pose, "position") is not JsonElement p ||
                    Obj(pose, "orientation") is not JsonElement o) return;
                _store.PosX = Num(p, "x") ?? 0;
                _store.PosY = Num(p, "y") ?? 0;
                _store.PosZ = Num(p, "z") ?? 0;
                var e = MathUtil.QuatToEuler(Num(o, "x") ?? 0, Num(o, "y") ?? 0, Num(o, "z") ?? 0, Num(o, "w") ?? 1);
                _store.Roll = e.Roll; _store.Pitch = e.Pitch; _store.Yaw = e.Yaw;
                _store.PushTrail(_store.PosX, _store.PosY);
                TelemetryStore.PushChart(_store.ChartAlt, _store.PosZ);
                _store.Touch();
                break;
            }
            case "/mavros/mavros/velocity_local":
            {
                if (Obj(msg, "twist") is not JsonElement twist ||
                    Obj(twist, "linear") is not JsonElement l) return;
                _store.Vx = Num(l, "x") ?? 0;
                _store.Vy = Num(l, "y") ?? 0;
                _store.Vz = Num(l, "z") ?? 0;
                TelemetryStore.PushChart(_store.ChartSpd, _store.SpeedH);
                _store.Touch();
                break;
            }
            case "/mavros/mavros/data":
            {
                if (Obj(msg, "angular_velocity") is not JsonElement av ||
                    Obj(msg, "linear_acceleration") is not JsonElement la) return;
                double gx = Num(av, "x") ?? 0, gy = Num(av, "y") ?? 0, gz = Num(av, "z") ?? 0;
                double ax = Num(la, "x") ?? 0, ay = Num(la, "y") ?? 0, az = Num(la, "z") ?? 0;
                _store.GyroMag = Math.Sqrt(gx * gx + gy * gy + gz * gz) * 180 / Math.PI;
                _store.AccelMag = Math.Sqrt(ax * ax + ay * ay + az * az);
                break;
            }
            default:
                return;
        }
        _store.NotifyUpdated();
    }
}
